libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,881 @@
+const MAX_ERASE_CYCLE: usize = 65535;
+
+/// Maximum page index.
+///
+/// Thus the maximum number of pages in one more than this number.
+const MAX_PAGE_INDEX: usize = 63;
+
+/// Maximum number of keys.
+const MAX_KEY_INDEX: usize = 4095;
+
+/// Maximum length in bytes of a user payload.
+const MAX_VALUE_LEN: usize = 1023;
+
+/// Maximum number of updates per transaction.
+const MAX_UPDATES: usize = 31;
+
+/// Maximum number of words per virtual page.
+const MAX_VIRT_PAGE_SIZE: usize = div_ceil(MAX_PAGE_SIZE, WORD_SIZE) - CONTENT_WORD;
+
+/// Word with all bits set to one.
+const ERASED_WORD: u32 = 0xffffffff;
+
+/// Helpers for a given storage configuration.
+#[derive(Clone, Debug)]
+pub struct Format {
+    /// The size in bytes of a page in the storage.
+    page_size: usize,
+
+    /// The number of pages in the storage.
+    num_pages: usize,
+
+    /// The maximum number of times a page can be erased.
+    max_page_erases: usize,
+}
+
+impl Format {
+    /// Extracts the format from a storage.
+    ///
+    /// Returns `None` if the storage is not [supported].
+    ///
+    /// [supported]: struct.Format.html#method.is_storage_supported
+    pub fn new<S: Storage>(storage: &S) -> Option<Format> {
+        if Format::is_storage_supported(storage) {
+            Some(Format {
+                page_size: storage.page_size(),
+                num_pages: storage.num_pages(),
+                max_page_erases: storage.max_page_erases(),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Returns whether a storage is supported.
+    ///
+    /// A storage is supported if the following conditions hold:
+    /// - The size of a word is 4 bytes.
+    /// - The size of a word evenly divides the size of a page.
+    /// - A page contains at least 8 words.
+    /// - A page contains at most [`MAX_PAGE_SIZE`] bytes.
+    /// - There is at least 3 pages.
+    /// - There is at most [`MAX_PAGE_INDEX`]` + 1` pages.
+    /// - A word can be written at least twice between erase cycles.
+    /// - A page can be erased at most [`MAX_ERASE_CYCLE`] times.
+    ///
+    /// [`MAX_PAGE_SIZE`]: constant.MAX_PAGE_SIZE.html
+    /// [`MAX_PAGE_INDEX`]: constant.MAX_PAGE_INDEX.html
+    /// [`MAX_ERASE_CYCLE`]: constant.MAX_ERASE_CYCLE.html
+    fn is_storage_supported<S: Storage>(storage: &S) -> bool {
+        let word_size = storage.word_size();
+        let page_size = storage.page_size();
+        let num_pages = storage.num_pages();
+        let max_word_writes = storage.max_word_writes();
+        let max_page_erases = storage.max_page_erases();
+        word_size == 4
+            && page_size % word_size == 0
+            && (8 * word_size <= page_size && page_size <= MAX_PAGE_SIZE)
+            && (3 <= num_pages && num_pages <= MAX_PAGE_INDEX + 1)
+            && max_word_writes >= 2
+            && max_page_erases <= MAX_ERASE_CYCLE
+    }
+
+    /// The size of a word in bytes.
+    pub fn word_size(&self) -> usize {
```

Comment:
I don't really understand why this is a function if the value is fixed to a constant. I'd understand it if this was implementing a trait that has a `word_size()` function, but here it's just a free function of the `Format`. Likewise for `max_key()` and `max_updates()`.

Reply:
Essentially 2 reasons:
1. The general design extends to different word sizes. The store is parametric over the word size (as is the storage interface). In particular the word size could be dynamic (useful for fuzzing different types of storage with the same code or having unit tests without using features).
2. It's more convenient to have a single object providing all the configuration parameters, in particular for a user to write `self.format.word_size()` instead `WORD_SIZE` or `format::WORD_SIZE`. All configuration parameters are of the form `self.format`.

The trait that implements `word_size` is `Storage`. All this library is parametric over `Storage` and as such parametric over `word_size`. It's just that the current implementation was optimized for `word_size == 4` only. But if we get new boards we could adapt the code to support them.

Reply:
> Essentially 2 reasons:
> 
> 1. The general design extends to different word sizes.

You mentioned elsewhere that no other word size than 4 is planned.

> The store is parametric over the word size (as is the storage interface). In particular the word size could be dynamic (useful for fuzzing different types of storage with the same code or having unit tests without using features).

If that was really a concern, const generics could also be used for such "dynamic" scenarios (where the word size is actually fixed for a given instance). Even though const generics are still an unstable feature, we already use a nightly compiler, which should support that. But anyway, I don't think the extra complexity of being able to check other word sizes than 4 is worth it at the moment.

> 2. It's more convenient to have a single object providing all the configuration parameters, in particular for a user to write `self.format.word_size()` instead `WORD_SIZE` or `format::WORD_SIZE`. All configuration parameters are of the form `self.format`.

This suggests that the word size is dynamic, when it's not. It's actually much clearer to have `self.format.foo()` for things that are really a dynamic property of the `Format` struct (i.e. change from one instance of `Format` to the other) and `Format::WORD_SIZE` for constants.

> The trait that implements `word_size` is `Storage`.

Then why does the `is_storage_supported` function above call `let word_size = storage.word_size();` but here we use `WORD_SIZE` instead of `self.word_size`?

> All this library is parametric over `Storage` and as such parametric over `word_size`. It's just that the current implementation was optimized for `word_size == 4` only. But if we get new boards we could adapt the code to support them.

I mean, either other word sizes aren't supported (as you suggested elsewhere) or they are. I just don't see the point of half supporting various sizes, half hard-coding "word_size == 4".

If we get new boards that don't use word-size == 4 (which is unlikely in the foreseeable future), we could adapt the code regardless. But if we don't get such new boards, half-supporting these hypothetical boards is unnecessary extra complexity.

Reply:
> You mentioned elsewhere that no other word size than 4 is planned.

It depends on the meaning of planned:
1. No other word size than 4 are planned in the sense that they are not planned to be done with our current knowledge (i.e. this series of PRs).
2. Other word size than 4 are planned in the sense that they are supported in case we have to support a board we currently don't know about.

Having the design parametric over the word size permits to not have to redo all the design work each time we have a new board. But only do some adaptation and tuning work.

> This suggests that the word size is dynamic, when it's not.

But it could be. I prefer to not hard-code whether it's dynamic or static in the syntax, but keep it at the semantics level only, such that we can easily change our mind as things evolve without having to update all the syntax.

> Then why does the is_storage_supported function above call let word_size = storage.word_size(); but here we use WORD_SIZE instead of self.word_size?

This is just to keep the store RAM usage at 3 words instead of 4. As long as we only support 32-bits words, we don't need to store it in the Format. This should be the only non-constant place where we use `WORD_SIZE`. The rest should use `word_size()`.

> half-supporting these hypothetical boards is unnecessary extra complexity

Could you describe what complexity you are talking about? The only thing I see is writing `word_size()` instead of `WORD_SIZE` which is not complexity.

Reply:
> Could you describe what complexity you are talking about? The only thing I see is writing `word_size()` instead of `WORD_SIZE` which is not complexity.

This function is extra lines of code, which is complexity. The `Storage` also has a `word_size()` function, which is extra code.

Sure, it's not a lot of complexity, but I'd argue that extending the current design to support other word sizes would also not be so much extra work (because the current design still has a concept of `WORD_SIZE`).

I also don't see any plans to support 64-bit based platforms in Tock for the time being, whereas [const generics](https://github.com/rust-lang/rust/issues/44580) are making progress in Rust, and would ultimately be a better fit than functions to represent the word size - for a given board there is only one possible word size known at compile time.

So relying on a fixed `const WORD_SIZE` everywhere is already minimizing the potential extra work to support future boards, assuming const generics will be stable enough in Rust by then (we already rely on the nightly compiler anyway). I don't see a compelling argument for representing the word size by a function for the time being.

> This is just to keep the store RAM usage at 3 words instead of 4.

Does this optimization matter? I don't see the code possibly using various instances of the `Format` so the RAM impact is at most a single word once. But the compiler may well realize that the word size is a constant and optimize it away.

> As long as we only support 32-bits words, we don't need to store it in the Format. This should be the only non-constant place where we use `WORD_SIZE`. The rest should use `word_size()`.

So the current implementation only supports `WORD_SIZE == 4`, and lifting this restriction wouldn't be only about removing the `word_size == 4` check in `is_storage_supported`. In that sense, other word sizes are only half-supported, which I think can become more confusing than not trying to support them at all - because someone trying to extend the current implementation to other word sizes would have to realize that.

In the end, we have to decide: either potentially "wasting" 1 word of RAM is fine and the `Format` should be fully dynamic on the word size, or this 1 word of RAM is unacceptable cost and the `Format` should be specialized for the only word size that matters, i.e. `WORD_SIZE == 4`.

Reply:
Regarding const generics, that's indeed the long-term plan. But for now the design is not yet clear due to lack of explicit use-cases. So there is no need to invest efforts in that yet.

Regarding support for other word sizes, it's not only about implementation, it's more importantly about design. The design and proofs support other word sizes. And that's the most complicated part of this library. The implementation is trivial. And this was actually a deliberate choice to keep the implementation trivial, because we can always optimize later as we see need and it permits to have more confidence in its correctness. Proving the correctness of a parametric algorithm is simpler than one which is instantiated.

> In the end, we have to decide: either potentially "wasting" 1 word of RAM is fine and the Format should be fully dynamic on the word size, or this 1 word of RAM is unacceptable cost and the Format should be specialized for the only word size that matters, i.e. WORD_SIZE == 4.

Then we're all good because the current implementation follows the second option. The Format is specialized for `WORD_SIZE == 4`. It's actually the definition of `WORD_SIZE`. Resolving the conversation.